Add support for configuring proxy scheme in S3 client settings and EC2 discovery plugin

heissanjay opened the pull request:
This PR introduces improvements to the S3 client configuration enables the specification of proxy schemes, providing more flexibility for users operating Elasticsearch in environments with specific proxy requirements.


Closes #101873

heissanjay committed:
Add support for configuring proxy scheme in S3 client settings

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

kingherc reviewed (commented):
Thanks @heissanjay ! This looks good to me. Can you:

* Add the new setting in the documentation page at `docs/reference/snapshot-restore/repository-s3.asciidoc` under `proxy.port`?
* Add a `docs/changelog/102495.yaml` file (see other similar files) summarizing this PR
* I see also that the `plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/Ec2ClientSettings.java` settings (docs at https://www.elastic.co/guide/en/elasticsearch/plugins/current/discovery-ec2-usage.html) also have the same missing proxy scheme. And the code looks very similar. I wonder if you could also introduce a similar setting there in this PR? Or, create a similar github issue for the discovery plugin that'll be handled in a different PR.

heissanjay committed:
docs: Add reference docs for S3 proxy scheme setting and Add changelog

kingherc commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3ClientSettings.java:
```diff
@@ -96,6 +96,13 @@ final class S3ClientSettings {
+    static final Setting.AffixSetting<Protocol> PROXY_SCHEME_SETTING = Setting.affixKeySetting(
+        PREFIX,
+        "proxy.scheme",
+        key -> new Setting<>(key, "https", s -> Protocol.valueOf(s.toUpperCase(Locale.ROOT)), Property.NodeScope)
```
Good catch! The ticket threw me off, because it says `Default seems to be https` but indeed it's http for the proxy protocol.

heissanjay replied:
@DaveCTurner @kingherc 

Thanks for highlighting the discussion around the default value for `proxy.scheme`.
After reviewing the current state and considering the feedback, I'm inclined to propose a change to set the default value of `proxy.scheme` to "http" in alignment with the observed behavior.

Would you be open to this adjustment, or is there any specific consideration or alternative preference that we should take into account before making this change? I 

Thanks!

kingherc replied:
@heissanjay yes please change the PR to make `http` the default for the scheme. Otherwise it would break current deployments when they upgrade to the latest version with this code (they were using the default `http` of S3 client, and after the upgrade it would try to use `https` and potentially fail).

kingherc commented on docs/reference/snapshot-restore/repository-s3.asciidoc:
```diff
@@ -133,6 +133,12 @@ settings belong in the `elasticsearch.yml` file.
+`proxy.scheme`::
+
+    The scheme to use for the proxy connection to S3. Valid values are either `http` or `https`.
+    Defaults to `https`. This setting allows to specify the protocol used for communication with the
```
Change to `http`.

kingherc commented on modules/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3ClientSettingsTests.java:
```diff
@@ -37,6 +37,7 @@ public void testThereIsADefaultClientByDefault() {
         assertThat(defaultSettings.protocol, is(Protocol.HTTPS));
         assertThat(defaultSettings.proxyHost, is(emptyString()));
         assertThat(defaultSettings.proxyPort, is(80));
+        assertThat(defaultSettings.proxyScheme, is(Protocol.HTTPS));
```
`HTTP`

heissanjay replied:
Sure @kingherc, I will update that. Please confirm this goes same for `plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/Ec2ClientSettings.java`. Thanks!!

kingherc replied:
Yes @heissanjay , both use ultimately https://github.com/aws/aws-sdk-java/blob/004bb46aa8b3912f172a1240c3f622336e167921/aws-java-sdk-core/src/main/java/com/amazonaws/ClientConfiguration.java#L171 which mentions a default value of HTTP for the proxy

heissanjay committed:
set default `proxy.scheme` to http

heissanjay committed:
Modified summary to include changes in EC2 discovery plugin

heissanjay committed:
Add proxy scheme configuration for EC2 discovery plugin

heissanjay commented:
@kingherc Thanks for your review and feedback! I've addressed the comments and made the necessary updates. The default value for `proxy.scheme` is now set to `http` to avoid potential issues during upgrades for existing deployments. I've also included the changes for the EC2 discovery plugin. Please review the updates, and let me know if any further adjustments are needed. Appreciate your collaboration!

kingherc commented:
@elasticmachine test this please

kingherc commented:
buildkite test this

kingherc commented:
Hi @heissanjay ,

- Can you fix the following errors?

```
AwsEc2ServiceImplTests.java:122: error: method launchAWSConfigurationTest in class AwsEc2ServiceImplTests cannot be applied to given types;
        launchAWSConfigurationTest(Settings.EMPTY, Protocol.HTTPS, null, -1, null, null, ClientConfiguration.DEFAULT_SOCKET_TIMEOUT);
        ^
  required: Settings,Protocol,String,int,Protocol,String,String,int
  found:    Settings,Protocol,<null>,int,<null>,<null>,int
  reason: actual and formal argument lists differ in length
/AwsEc2ServiceImplTests.java:137: error: incompatible types: String cannot be converted to Protocol
        launchAWSConfigurationTest(settings, Protocol.HTTP, "aws_proxy_host", 8080, "http",  "aws_proxy_username", "aws_proxy_password", 10000);
                                                                                    ^
```

- Can you run `./gradlew spotlessApply` (see [CONTRIBUTING](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md#java-language-formatting-guidelines)) to fix some code style issues and push to the PR again?
- Also please run the `testThereIsADefaultClientByDefault` , `testAWSConfigurationWithAwsSettings`, and `Ec2DiscoveryPluginTests.java` tests (the ones this PR modifies) manually to ensure they work.

Afterwards I'll re-test your PR before approving.

heissanjay committed:
fix Type error in the Test cases and code styling

heissanjay commented:
@kingherc I've made the necessary fixes to address the issues you mentioned. I've also run `./gradlew spotlessApply` to ensure code style compliance. Could you please trigger the test again? Thanks.

onlysavior commented:
您好，来信已收到，回尽快给您回复，祝身体健康

kingherc commented:
buildkite test this

kingherc reviewed (approved):
Tests passed. LGTM. Merging it.
